activerecord/lib/active_record/relation/where_clause.rb

```diff
@@ -80,8 +80,14 @@ def invert
         WhereClause.new(inverted_predicates, binds)
       end
 
+      def freeze
+        # This method memoizes, which will blow up on a frozen object
+        referenced_columns
+        super
+      end
+
       def self.empty
-        new([], [])
+        @empty ||= new([], []).freeze
```

Comment:
This object exposes no mutable methods in its API. I see no reason for us to freeze it, as we should be free to memoize whenever we want without having to worry about being frozen.